### PR TITLE
MINIFICPP-2470 Add SSL and mTLS authentication support to CouchbaseClusterService

### PR DESCRIPTION
**Pull request by lordgamez**
- Add mTLS authentication option with linked SSLContextService of CouchbaseClusterService
- Add support for adding CA certificate with linked SSLContextService to use SSL communication when using username/password authentication
- Fix enabling linked services of a controller service before enabling the controller service itself

https://issues.apache.org/jira/browse/MINIFICPP-2470

Depends on https://github.com/apache/nifi-minifi-cpp/pull/1881

-------------------------------------
Thank you for submitting a contribution to Apache NiFi - MiNiFi C++.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced
     in the commit message?

- [ ] Does your PR title start with MINIFICPP-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically main)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)?
- [ ] If applicable, have you updated the LICENSE file?
- [ ] If applicable, have you updated the NOTICE file?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check GitHub Actions CI results for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Review by adamdebreceni — APPROVED** [review]

File: extensions/couchbase/controllerservices/CouchbaseClusterService.cpp [hunk 1]
```diff
@@ -26,6 +26,50 @@
...
+    : connection_string_(std::move(connection_string)), logger_(logger), cluster_options_(buildClusterOptions(std::move(username), std::move(password), ssl_context_service)) {
+}
+
+::couchbase::cluster_options CouchbaseClient::buildClusterOptions(std::string username, std::string password, minifi::controllers::SSLContextService* ssl_context_service) {
+  if (username.empty() && (!ssl_context_service || (ssl_context_service && ssl_context_service->getCertificateFile().empty()))) {
+    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Neither username and password nor SSLContextService is provided for Couchbase authentication");
+  }
+
+  if (!username.empty() && ssl_context_service && !ssl_context_service->getCertificateFile().empty()) {
+    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Username and password authentication or mTLS authentication using certificate defined in SSLConextService "
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply extensions/couchbase/controllerservices/CouchbaseClusterService.cpp]: Good catch, updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/0abf6fbc2bc0d13c0c28fa40a01da816911a12b8

File: extensions/couchbase/controllerservices/CouchbaseClusterService.cpp [hunk 2]
```diff
@@ -159,11 +202,24 @@ void CouchbaseClusterService::onEnable() {
...
   std::string password;
   getProperty(UserPassword, password);
-  if (connection_string.empty() || username.empty() || password.empty()) {
-    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Missing connection string, username or password");
+  if (connection_string.empty()) {
+    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Missing connection string");
+  }
+
+  if ((username.empty() || password.empty()) && linked_services_.empty()) {
+    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Missing username and password or SSLConextService as a linked service");
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply extensions/couchbase/controllerservices/CouchbaseClusterService.cpp]: Updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/0abf6fbc2bc0d13c0c28fa40a01da816911a12b8

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py [hunk 3]
```diff
@@ -12,19 +12,72 @@
...
+
+        self.couchbase_cert_file = tempfile.NamedTemporaryFile(delete=False)
+        self.couchbase_cert_file.write(OpenSSL.crypto.dump_certificate(type=OpenSSL.crypto.FILETYPE_PEM, cert=couchbase_cert))
+        self.couchbase_cert_file.close()
+        os.chmod(self.couchbase_cert_file.name, 0o666)
+
+        self.couchbase_key_file = tempfile.NamedTemporaryFile(delete=False)
+        self.couchbase_key_file.write(OpenSSL.crypto.dump_privatekey(type=OpenSSL.crypto.FILETYPE_PEM, pkey=couchbase_key))
+        self.couchbase_key_file.close()
+        os.chmod(self.couchbase_key_file.name, 0o666)
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py:45]: The ssl flag should not be needed anymore, as the couchbase server is configured to serve both ssl and non-ssl clients, removed the ssl flag in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/c3519585ea81b9d4c89383e5138be9d37980a99e

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py [hunk 4]
```diff
@@ -12,19 +12,72 @@
...
 
     def get_startup_finished_log_entry(self):
         # after startup the logs are only available in the container, only this message is shown
         return "logs available in"
 
+    @retry_check(12, 5)
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py]: Updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/c3519585ea81b9d4c89383e5138be9d37980a99e

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py [hunk 5]
```diff
@@ -12,19 +12,72 @@
...
+        if code != 0:
+            logging.error(f"Failed to run command '{command}', returned error code: {code}")
+            return False
+        return True
+
+    def _run_couchbase_cli_commands(self, commands):
+        for command in commands:
+            if not self._run_couchbase_cli_command(command):
+                return False
+        return True
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py]: Updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/c3519585ea81b9d4c89383e5138be9d37980a99e

File: docker/test/integration/features/steps/steps.py [hunk 6]
```diff
@@ -1384,6 +1384,34 @@ def step_impl(context, service_name):
     container.add_controller(couchbase_cluster_controller_service)
 
 
+@given("a CouchbaseClusterService is setup up with SSL connection with the name \"{service_name}\"")
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/features/steps/steps.py]: Updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/c3519585ea81b9d4c89383e5138be9d37980a99e

File: extensions/couchbase/controllerservices/CouchbaseClusterService.cpp [hunk 7]
```diff
@@ -50,6 +50,50 @@ CouchbaseErrorType getErrorType(const std::error_code& error_code) {
...
+}
+
+::couchbase::cluster_options CouchbaseClient::buildClusterOptions(std::string username, std::string password, minifi::controllers::SSLContextService* ssl_context_service) {
+  if (username.empty() && (!ssl_context_service || (ssl_context_service && ssl_context_service->getCertificateFile().empty()))) {
+    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Neither username and password nor SSLContextService is provided for Couchbase authentication");
+  }
+
+  if (!username.empty() && ssl_context_service && !ssl_context_service->getCertificateFile().empty()) {
+    throw minifi::Exception(ExceptionType::PROCESS_SCHEDULE_EXCEPTION, "Username and password authentication or mTLS authentication using certificate defined in SSLContextService "
+      "linked service should be provided exclusively for Couchbase");
```
[the remark this answers is not in the document]
Comment (lordgamez) [reply extensions/couchbase/controllerservices/CouchbaseClusterService.cpp]: Updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/c3519585ea81b9d4c89383e5138be9d37980a99e

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (fgerlits) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py:45]: the `ssl=False` parameter could be removed, as well

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (fgerlits) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py]: can you make the same change at `_load_couchbase_certs()`, too, please?

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py:45]: Good point, updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/579e378d6343930fbd22ecc2e8852df74314e815

File: docker/test/integration/cluster/containers/CouchbaseServerContainer.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (lordgamez) [reply docker/test/integration/cluster/containers/CouchbaseServerContainer.py]: Updated in https://github.com/apache/nifi-minifi-cpp/pull/1882/commits/579e378d6343930fbd22ecc2e8852df74314e815

**Review by fgerlits — APPROVED** [review]

**Review by szaszm — APPROVED** [review]

**Pull request closed without merging**